### ISSUE TITLE
Should be able to set type on functions

### ISSUE THREAD

**Issue by zavr-1** [google/closure-compiler#3528]
TypeScript allows to set a type on a function 
![image](https://user-images.githubusercontent.com/21156791/71304856-22fd6d80-23dd-11ea-860f-903d352576a7.png)
CC should allow that too 
This will allow to define a function as a type in one file and then use it in implementation. 
I understand everyone's very conservative here I remember I opened a similar issue I got an explanation that functions should not be annotated with `@type` but I think they should. This way I can create interfaces for functions, import them in implementations and have type checking doe by the compiler. Otherwise, I have manually add the parameters to the function which I don't want to do because it's wasting time since I already have function's types in a separate place. TypeScript in vs code does it properly as you can see on the screenshot.

**Comment by zavr-1** [comment]
OK so i started to assume that it's not possible to set types like that because my typechecking wasn't done when trying to set `app.anything` because i expected an error there saying that this property wasn't define it app's constructor but it didn't. when i removed the type and returned params the compiler still didn't say anything so i don't know what the problem was. the issue can then be converted into the following one 
```js
/**
 * @type {_idio.session}
 */
export default function(app, opts = {}) {
}
```
```
src/index.js:31: WARNING - [JSC_MISPLACED_ANNOTATION] Misplaced type annotation. Type annotations are not allowed here. Are you missing parentheses?
export default function(app, opts = {}) {
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
```
i.e., not being able to annotate type on default export which shouldn't be an obstacle.
